### PR TITLE
[Merged by Bors] - feat: port Topology.UniformSpace.UniformConvergenceTopology

### PR DESCRIPTION
**Pull request by mcdoll**
---

- [x] depends on: #2092

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#2092
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#2092~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by RemyDegenne** [comment]
It looks like most of the issues in this file come from the fact that the lean 3 code used a lot the fact that `α →ᵤ β` is definitionally equal to `α → β`, but using one for another won't work in lean 4.
I think that the remaining failures in `rw` near the end of the file are due to this. I don't know about the other errors.

File: Mathlib/Order/Filter/Bases.lean [hunk 1]
```diff
@@ -466,7 +466,7 @@ theorem HasBasis.le_iff (hl : l.HasBasis p s) : l ≤ l' ↔ ∀ t ∈ l', ∃ i
 theorem HasBasis.le_basis_iff (hl : l.HasBasis p s) (hl' : l'.HasBasis p' s') :
     l ≤ l' ↔ ∀ i', p' i' → ∃ i, p i ∧ s i ⊆ s' i' := by
   simp only [hl'.ge_iff, hl.mem_iff]
-#align filter.has_basis.le_basis_iff Filter.HasBasis.le_basis_iffₓ
+#align filter.has_basis.le_basis_iff Filter.HasBasis.le_basis_iff
```
Comment (jcommelin) [inline Mathlib/Order/Filter/Bases.lean:469]: Why this change? Presumably the `ₓ` was added for a good reason.

Comment (RemyDegenne) [reply Mathlib/Order/Filter/Bases.lean:469]: Perhaps there is a good reason indeed. I don't know the meaning of that x.
The issue that prompted the change is that in the main file of this PR there was an error because the lean 3 lemma `le_basis_iff` was translated to `le_basis_iffₓ`, which is apparently not a lemma that exists? The lemma above the align is `le_basis_iff` without the x and using that lemma works, so I thought the align was a mistake?

Comment (jcommelin) [reply Mathlib/Order/Filter/Bases.lean:469]: Aha. That sounds like a bug in mathport. The `ₓ` in an align statement signifies that the translation isn't exactly faithful. Most cases I've seen so far are because of reordered universes.
But it should not appear in places where the lemma is used...
```suggestion
#align filter.has_basis.le_basis_iff Filter.HasBasis.le_basis_iffₓ
```

File: Mathlib/Topology/UniformSpace/UniformConvergenceTopology.lean [hunk 2]
```diff
@@ -0,0 +1,942 @@
...
+  of the form `S(V) := {(f, g) | ∀ x : α, (f x, g x) ∈ V}` for some `V : Set (β × β)`
+* `uniform_fun.uniform_space`: uniform structure of uniform convergence. This is the
+  `UniformSpace` on `α →ᵤ β` whose uniformity is generated by the sets `S(V)` for `V ∈ 𝓤 β`.
+  We will denote this uniform space as `𝒰(α, β, uβ)`, both in the comments and as a local notation
+  in the Lean code, where `uβ` is the uniform space structure on `β`.
+  This is declared as an instance on `α →ᵤ β`.
+* `uniform_on_fun.uniform_space`: uniform structure of `𝔖`-convergence, where
+  `𝔖 : Set (Set α)`. This is the infimum, for `S ∈ 𝔖`, of the pullback of `𝒰 S β` by the map of
+  restriction to `S`. We will denote it `𝒱(α, β, 𝔖, uβ)`, where `uβ` is the uniform space structure
+  on `β`.
```
Comment (dupuisf) [inline Mathlib/Topology/UniformSpace/UniformConvergenceTopology.lean:47]:
```suggestion
* `UniformOnFun.uniformSpace`: uniform structure of `𝔖`-convergence, where
  `𝔖 : Set (Set α)`. This is the infimum, for `S ∈ 𝔖`, of the pullback of `𝒰 S β` by the map of
  restriction to `S`. We will denote it `𝒱(α, β, 𝔖, uβ)`, where `uβ` is the uniform space structure
  on `β`.
```

File: Mathlib/Topology/UniformSpace/UniformConvergenceTopology.lean [hunk 3]
```diff
@@ -0,0 +1,942 @@
...
+scoped[UniformConvergence] notation:25 α " →ᵤ " β:0 => UniformFun α β
+
+-- mathport name: «expr →ᵤ[ ] »
+scoped[UniformConvergence] notation:25 α " →ᵤ[" 𝔖 "] " β:0 => UniformOnFun α β 𝔖
+
+-- mathport name: «exprλᵘ , »
+-- scoped[UniformConvergence] notation3"λᵘ "(...)", "r:(scoped p => UniformFun.ofFun p) => r
+
+-- mathport name: «exprλᵘ[ ] , »
+-- scoped[UniformConvergence] notation3"λᵘ["𝔖"] "(...)", "r:(scoped p => UniformFun.ofFun p) => r
```
Comment (dupuisf) [inline Mathlib/Topology/UniformSpace/UniformConvergenceTopology.lean:170]:
```suggestion
scoped[UniformConvergence] notation:25 α " →ᵤ " β:0 => UniformFun α β

scoped[UniformConvergence] notation:25 α " →ᵤ[" 𝔖 "] " β:0 => UniformOnFun α β 𝔖

-- scoped[UniformConvergence] notation3"λᵘ "(...)", "r:(scoped p => UniformFun.ofFun p) => r

-- scoped[UniformConvergence] notation3"λᵘ["𝔖"] "(...)", "r:(scoped p => UniformFun.ofFun p) => r
```

File: Mathlib/Topology/UniformSpace/UniformConvergenceTopology.lean [hunk 4]
```diff
@@ -0,0 +1,942 @@
...
+/-- Basis sets for the uniformity of uniform convergence: `gen α β V` is the set of pairs `(f, g)`
+of functions `α →ᵤ β` such that `∀ x, (f x, g x) ∈ V`. -/
+protected def gen (V : Set (β × β)) : Set ((α →ᵤ β) × (α →ᵤ β)) :=
+  { uv : (α →ᵤ β) × (α →ᵤ β) | ∀ x, (uv.1 x, uv.2 x) ∈ V }
+#align uniform_fun.gen UniformFun.gen
+
+/-- If `𝓕` is a filter on `β × β`, then the set of all `uniform_convergence.gen α β V` for
+`V ∈ 𝓕` is a filter basis on `(α →ᵤ β) × (α →ᵤ β)`. This will only be applied to `𝓕 = 𝓤 β` when
+`β` is equipped with a `UniformSpace` structure, but it is useful to define it for any filter in
+order to be able to state that it has a lower adjoint (see `uniform_convergence.gc`). -/
```
Comment (dupuisf) [inline Mathlib/Topology/UniformSpace/UniformConvergenceTopology.lean:221]: What are `uniform_convergence.gen` and `uniform_convergence.gc`? They don't exist in mathlib3 according to the docs -- was there a name change at some point...? Perhaps @ADedecker knows?

Comment (mcdoll) [reply Mathlib/Topology/UniformSpace/UniformConvergenceTopology.lean:217]: I would guess that it is supposed to be `UniformFun.gen` and `UniformFun.gc`.

File: Mathlib/Topology/UniformSpace/UniformConvergenceTopology.lean [hunk 5]
```diff
@@ -0,0 +1,939 @@
...
+
+/-- Uniform structure of `𝔖`-convergence, i.e uniform convergence on the elements of `𝔖`,
+declared as an instance on `α →ᵤ[𝔖] β`. It is defined as the infimum, for `S ∈ 𝔖`, of the pullback
+by `S.restrict`, the map of restriction to `S`, of the uniform structure `𝒰(s, β, uβ)` on
+`↥S →ᵤ β`. We will denote it `𝒱(α, β, 𝔖, uβ)`, where `uβ` is the uniform structure on `β`. -/
+instance uniformSpace : UniformSpace (α →ᵤ[𝔖] β) :=
+  ⨅ (s : Set α) (_hs : s ∈ 𝔖), UniformSpace.comap s.restrict 𝒰(s, β, _)
+
+-- mathport name: «expr𝒱( , , , )»
+local notation "𝒱(" α ", " β ", " 𝔖 ", " u ")" => @UniformOnFun.uniformSpace α β u 𝔖
```
Comment (dupuisf) [inline Mathlib/Topology/UniformSpace/UniformConvergenceTopology.lean:619]:
```suggestion
local notation "𝒱(" α ", " β ", " 𝔖 ", " u ")" => @UniformOnFun.uniformSpace α β u 𝔖
```

**Review by dupuisf — COMMENTED** [review]
Thanks!

bors r+

File: Mathlib/Topology/UniformSpace/UniformConvergenceTopology.lean (hunk 4 above)
Comment (ADedecker) [reply Mathlib/Topology/UniformSpace/UniformConvergenceTopology.lean]: Indeed I forgot to change the docstring
